support/sys-socket.c

```diff
@@ -123,6 +173,290 @@ Mono_Posix_Syscall_setsockopt_linger (int socket, int level, int option_name, st
 	return setsockopt (socket, level, option_name, &ling, sizeof (struct linger));
 }
 
+static int
+get_addrlen (struct Mono_Posix_Sockaddr* address, socklen_t* addrlen)
+{
+	if (!address) {
+		*addrlen = 0;
+		return 0;
+	}
+
+	switch (address->type) {
+	case Mono_Posix_SockaddrType_SockaddrStorage:
+		mph_return_if_socklen_t_overflow (((struct Mono_Posix__SockaddrDynamic*) address)->len);
+		*addrlen = ((struct Mono_Posix__SockaddrDynamic*) address)->len;
+		return 0;
+	case Mono_Posix_SockaddrType_SockaddrUn:
+		mph_return_if_socklen_t_overflow (offsetof (struct sockaddr_un, sun_path) + ((struct Mono_Posix__SockaddrDynamic*) address)->len);
+		*addrlen = offsetof (struct sockaddr_un, sun_path) + ((struct Mono_Posix__SockaddrDynamic*) address)->len;
+		return 0;
+	case Mono_Posix_SockaddrType_Sockaddr: *addrlen = sizeof (struct sockaddr); return 0;
+	case Mono_Posix_SockaddrType_SockaddrIn: *addrlen = sizeof (struct sockaddr_in); return 0;
+	case Mono_Posix_SockaddrType_SockaddrIn6: *addrlen = sizeof (struct sockaddr_in6); return 0;
+	default:
+		*addrlen = 0;
+		errno = EINVAL;
+		return -1;
+	}
+}
+
+int
+Mono_Posix_Sockaddr_GetNativeSize (struct Mono_Posix_Sockaddr* address, gint64* size)
+{
+	socklen_t value;
+	int r;
+
+	r = get_addrlen (address, &value);
+	*size = value;
+	return r;
+}
+
+int
+Mono_Posix_FromSockaddr (struct Mono_Posix_Sockaddr* source, void* destination)
+{
+	if (!source)
+		return 0;
+
+	switch (source->type) {
+	case Mono_Posix_SockaddrType_SockaddrStorage:
+		// Do nothing, don't copy source->sa_family into addr->sa_family
+		return 0;
+
+	case Mono_Posix_SockaddrType_SockaddrUn:
+		memcpy (((struct sockaddr_un*) destination)->sun_path, ((struct Mono_Posix__SockaddrDynamic*) source)->data, ((struct Mono_Posix__SockaddrDynamic*) source)->len);
+		break;
+
+	case Mono_Posix_SockaddrType_Sockaddr:
+		break;
+
+	case Mono_Posix_SockaddrType_SockaddrIn:
+		if (Mono_Posix_FromSockaddrIn ((struct Mono_Posix_SockaddrIn*) source, (struct sockaddr_in*) destination) != 0)
+			return -1;
+		break;
+
+	case Mono_Posix_SockaddrType_SockaddrIn6:
+		if (Mono_Posix_FromSockaddrIn6 ((struct Mono_Posix_SockaddrIn6*) source, (struct sockaddr_in6*) destination) != 0)
+			return -1;
+		break;
+
+	default:
+		errno = EINVAL;
+		return -1;
+	}
+
+	int family;
+	if (Mono_Posix_FromUnixAddressFamily (source->sa_family, &family) != 0)
+		return -1;
+	((struct sockaddr*) destination)->sa_family = family;
+
+	return 0;
+}
+
+int
+Mono_Posix_ToSockaddr (void* source, gint64 size, struct Mono_Posix_Sockaddr* destination)
+{
+	struct Mono_Posix__SockaddrDynamic* destination_dyn;
+
+	if (!destination)
+		return 0;
+
+	switch (destination->type) {
+	case Mono_Posix_SockaddrType_Sockaddr:
+		if (size < offsetof (struct sockaddr, sa_family) + sizeof (sa_family_t)) {
+			errno = ENOBUFS;
+			return -1;
+		}
+		break;
+
+	case Mono_Posix_SockaddrType_SockaddrStorage:
+		destination_dyn = ((struct Mono_Posix__SockaddrDynamic*) destination);
```

Comment:
`SockaddrStorage` and `SockaddrUn` require a dynamic amount of memory, the other types always require the same amount.
The C# code will create a `_SockaddrDynamic` instance if the `SockaddrType` has the `MustBeWrapped` flag set. (And in case the C# code should forget to do that, the switch will fall through to the `default:` case.)